readers/apiserver/watchlist/internal/watchlist/relist/listers.go

```diff
@@ -0,0 +1,136 @@
+package relist
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/Azure/tattler/data"
+	"github.com/gostdlib/base/retry/exponential"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var backoff = exponential.Must(exponential.New())
+
+// genericLister is a function that lists resources of type T. It can be any of the common
+// client list functions that return a concrete type, such as *corev1.NamespaceList.
+// So for example, r.clientset.CoreV1().Namespaces().List is a genericLister[*corev1.NamespaceList].
+type genericLister[T runtime.Object] func(ctx context.Context, options metav1.ListOptions) (T, error)
+
+// adapter will turn any genericLister into a ListPage. A generic lister is any of the common client
+// list functions that return a concrete type, such as *corev1.NamespaceList. The returned ListPage
+// also will handle exponential backoff retries for retriable errors.
+func adapter[R runtime.Object](f genericLister[R]) listPage {
+	return func(ctx context.Context, options metav1.ListOptions) ([]data.Entry, string, error) {
+		var entries []data.Entry
+		var continueToken string
+		err := backoff.Retry(
+			ctx,
+			func(ctx context.Context, _ exponential.Record) error {
+				obj, err := f(ctx, options)
+				if err != nil {
+					return err
+				}
+				entries, err = listToEntries(obj)
+				if err != nil {
+					return err
+				}
+				continueToken, err = getContinue(obj)
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			exponential.WithMaxAttempts(10),
+		)
+		return entries, continueToken, err
+	}
+}
+
+// listToEntries converts a List() result to a slice of data.Entry.
+func listToEntries(listResult runtime.Object) ([]data.Entry, error) {
+	var events = make([]data.Entry, 0, 100)
+
+	if listResult == nil {
+		return nil, fmt.Errorf("listToEntries: listResult is nil: %w", exponential.ErrPermanent)
+	}
+
+	// Use reflection to extract Items field from any List type (PodList, NodeList, etc.)
+	listValue := reflect.ValueOf(listResult)
```

Comment:
are you not able to use something likee listResult.GetObjectKind().GroupVersionKind()

Reply:
I think in general I'm not sure why you're not using the existing k8s libraries for dealing with runtime.Object. It seems like it would be safer to use than reflection.

https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#UnstructuredList